core/chains/terra/types/chain_set.go

```diff
@@ -0,0 +1,194 @@
+type ChainSetOpts struct {
+	Config           coreconfig.GeneralConfig
+	Logger           logger.Logger
+	DB               *sqlx.DB
+	KeyStore         keystore.Terra
+	EventBroadcaster pg.EventBroadcaster
+	ORM              ORM
+}
+
+func (o ChainSetOpts) validate() (err error) {
+	required := func(s string) error {
+		return fmt.Errorf("%s is required", s)
+	}
+	if o.Config == nil {
+		err = multierr.Append(err, required("Config"))
+	}
+	if o.Logger == nil {
+		err = multierr.Append(err, required("Logger'"))
+	}
+	if o.DB == nil {
+		err = multierr.Append(err, required("DB"))
+	}
+	if o.KeyStore == nil {
+		err = multierr.Append(err, required("KeyStore"))
+	}
+	if o.EventBroadcaster == nil {
+		err = multierr.Append(err, required("EventBroadcaster"))
+	}
+	if o.ORM == nil {
+		err = multierr.Append(err, required("ORM"))
+	}
+	return
+}
+
+func (o ChainSetOpts) newChain(dbchain Chain) (*chain, error) {
+	return NewChain(o.DB, o.KeyStore, o.Config, o.EventBroadcaster, dbchain, o.Logger)
+}
+
+//go:generate mockery --name ChainSet --srcpkg github.com/smartcontractkit/chainlink-terra/pkg/terra --output ./mocks/ --case=underscore
+var _ terra.ChainSet = (*chainSet)(nil)
+
+type chainSet struct {
+	utils.StartStopOnce
+	opts     ChainSetOpts
+	chainsMu sync.RWMutex
+	chains   map[string]terra.Chain
+	lggr     logger.Logger
+}
+
+// NewChainSet returns a new chain set for opts.
+func NewChainSet(opts ChainSetOpts) (terra.ChainSet, error) {
+	if err := opts.validate(); err != nil {
+		return nil, err
+	}
+	dbchains, err := opts.ORM.EnabledChainsWithNodes()
+	if err != nil {
+		return nil, errors.Wrap(err, "error loading chains")
+	}
+	cs := chainSet{
+		opts:   opts,
+		chains: make(map[string]terra.Chain),
+		lggr:   opts.Logger.Named("ChainSet"),
+	}
+	for _, dbc := range dbchains {
+		var err2 error
+		cs.chains[dbc.ID], err2 = opts.newChain(dbc)
+		if err2 != nil {
+			err = multierr.Combine(err, err2)
+			continue
+		}
+	}
+	return &cs, err
+}
+
+func (c *chainSet) Chain(id string) (terra.Chain, error) {
+	if err := c.StartStopOnce.Ready(); err != nil {
+		return nil, err
+	}
+	c.chainsMu.RLock()
+	ch := c.chains[id]
+	c.chainsMu.RUnlock()
+	if ch != nil {
+		// Already known/started
+		return ch, nil
+	}
+
+	// Unknown/unstarted
+	c.chainsMu.Lock()
+	defer c.chainsMu.Unlock()
+
+	// Double check now that we have the lock, so we don't start an orphan.
+	if err := c.StartStopOnce.Ready(); err != nil {
+		return nil, err
+	}
+
+	ch = c.chains[id]
+	if ch != nil {
+		// Someone else beat us to it
+		return ch, nil
+	}
+
+	// Do we have nodes/config?
+	opts := c.opts
+	dbchain, err := opts.ORM.Chain(id)
+	if err != nil {
+		return nil, err
+	}
+
+	// Start it
+	tChain, err := opts.newChain(dbchain)
+	if err != nil {
+		return nil, err
+	}
+	err = tChain.Start()
+	if err != nil {
+		return nil, err
+	}
+	c.chains[id] = tChain
+	return tChain, nil
+}
+
+func (c *chainSet) Start() error {
+	//TODO if terra disabled, warn and return?
```

Comment:
We shouldnt even be able to get here I think if terra is disabled, but worthwhile adding a catch for it here anyway

Reply:
This mirrors the EVM version. If we are going to refactor here we should consider it as well.